File: code/query_string_parser.py
class QueryStringParser:
    def __init__(self, url):
        self.url = url
        a = url.find('?')
        if a != -1:
            q_str = url[a+1:len(url)]
            q_array = q_str.split('&')
            self.hash = {}
            for part in q_array:
                parts = part.split('=', 1)
                if parts[0] == None or parts[0] == '':
                    continue
                if parts[0] in self.hash:
                    if type(self.hash[parts[0]]) is list:
                        self.hash[parts[0]].append(parts[1])
                    else:
                        self.hash[parts[0]] = [self.hash[parts[0]], parts[1]]
                else:
                    if len(parts) == 1:
                        self.hash[parts[0]] = None
                    else:
                        self.hash[parts[0]] = parts[1]
        else:
            self.hash = {}

    def query_hash(self):
        results = {}
        for key, value in self.hash.iteritems():
            if type(value) is list:
                results[key] = value
            elif value:
                results[key] = value.replace('+', ' ')
                results[key] = results[key].replace('%26', '&')

                if key.find('+') != -1 or key.find('%26') != -1:
                    results[key.replace('+', ' ').replace('%26', '&')] = results[key]
                    del results[key]
            else:
                results[key] = value
        return results

File: code/js/src/query_string_parser.js
function QueryStringParser(url) {
    this.url = url;
    a = url.indexOf('?');
    if (a != -1) {
        q_str = url.substring(a+1, url.length);
        q_array = q_str.split('&')
        this.hash = {};
        debugger
        for (i = 0; i < q_array.length; i++) {
            part = q_array[i];
            parts = this.splitQueryStringEqual(part);
            if (parts[1] == undefined)
                parts[1] = null;

            if (parts[0] == null || parts[0] == '')
                continue;
            if (parts[0] in this.hash) {
                if (this.hash[parts[0]] instanceof Array)
                    this.hash[parts[0]].push(parts[1]);
                else
                    this.hash[parts[0]] = [this.hash[parts[0]], parts[1]]
            }
            else {
                this.hash[parts[0]] = parts[1];
            }
        }
    }
    else
        this.hash = {};
}

QueryStringParser.prototype.queryHash = function() {
    results = {}
    for (key in this.hash) {
        value = this.hash[key];
        if (value instanceof Array) {
            results[key] = value;
        }
        else if (value) {
            results[key] = value.replace(/\+/g, ' ');
            results[key] = results[key].replace(/%26/g, '&');

            if (key.indexOf('+') != -1 || key.indexOf('%26') != -1) {
                results[key.replace(/\+/g, ' ').replace(/%26/g, '&')] = results[key]
                delete results[key];
            }
        }
        else
            results[key] = value;
    }

    return results;
};

QueryStringParser.prototype.splitQueryStringEqual = function(part) {
    index = part.indexOf('=')
    if (index == -1)
        return [part];
    
    key = part.substring(0, index);
    value = part.substring(index + 1, part.length);
    return [key, value];
};
